# Proyecto de Analítica Predictiva para Ventas e Inventarios

## Samsung Innovation Campus - Hackathon
### Equipo: `Falta ";"`

### Integrantes
- Miguel Angel Alfonso Saavedra

### Metodología Implementada: ASUM-DM

# 1. Introducción y Contexto del Problema

Aqui se realizará la contextualización de la empresa y su problema para dar pie a la idea de lo que se quiere lograr con los datos.

### Contextualización

La empresa dispone de información histórica relacionada con las ventas e inventarios de sus productos distribuidos a diferentes clientes. Sin embargo, actualmente estos datos no están siendo aprovechados de forma estratégica para identificar patrones de comportamiento, apoyar la toma de decisiones o anticipar escenarios futuros del negocio.

El uso de analítica predictiva permitiría transformar la información histórica en conocimiento útil para las áreas operativas y comerciales. Por ejemplo, anticipar incrementos en la demanda de determinados productos permitiría ajustar oportunamente los niveles de inventario y reducir riesgos de desabastecimiento. De igual manera, identificar disminuciones en el volumen de compra de ciertos clientes facilitaría la implementación de estrategias comerciales preventivas.

Bajo este contexto, surge la necesidad de mejorar el aprovechamiento de la información disponible mediante herramientas analíticas que permitan comprender el comportamiento de las ventas, detectar patrones relevantes y estimar tendencias futuras.

Por esta razón, el área encargada plantea el siguiente problema de negocio:

`Actualmente la empresa cuenta con información histórica de ventas e inventarios, pero no dispone de herramientas analíticas predictivas que apoyen la planeación comercial y operativa. Esta limitación dificulta anticipar fluctuaciones en la demanda, optimizar los niveles de inventario y reducir errores en la estimación de ventas futuras.`

Con el fin de abordar esta problemática, a continuación se realizará un análisis de los datos históricos para identificar comportamientos relevantes y desarrollar modelos que permitan estimar tendencias futuras asociadas a las ventas y al inventario.

# 2. Comprensión del Negocio

Procedemos en esta sección a realizar el análisis del negocio en pro de explicar los objetivos de negocio para este proyecto. Adicionalmente se traducira las metas del negocio desde una perspectiva análitica, esto con el fin de relacionar ambas partes y generar las respectivas hipotesis a probar.


#### Negocio

Actualmente la empresa cuenta con información histórica relacionada con ventas e inventarios de productos comercializados a diferentes clientes. No obstante, esta información aún no está siendo utilizada de manera analítica para apoyar procesos de planeación, abastecimiento y toma de decisiones comerciales.

Debido a esta situación, surge la necesidad de analizar los datos disponibles con el propósito de identificar patrones relevantes, comprender el comportamiento de la demanda y desarrollar modelos predictivos que permitan estimar el comportamiento futuro de indicadores clave asociados a ventas e inventario.

A partir de este contexto, el proyecto busca construir una solución analítica que contribuya a mejorar la capacidad de respuesta del negocio frente a cambios en la demanda y posibles riesgos operativos.


*`Objetivos Estrátegicos`*

- Reducir el error en la estimación de ventas futuras.

- Mejorar la capacidad de anticipar cambios en la demanda de productos.

- Fortalecer la planeación de inventario y abastecimiento.

- Apoyar la toma de decisiones comerciales mediante análisis basados en datos.

---------------------------------------------------------------------------
*`Indicadores de Interés`*

Los principales indicadores que se desean analizar o estimar son:

- Ventas futuras por producto y cliente.

- Volumen esperado de ventas en próximos periodos.

- Productos con mayor o menor demanda esperada.

- Posibles riesgos de desabastecimiento según la demanda proyectada.

- Tendencias temporales en ventas e inventarios.

Adicionalmente, los resultados podrán presentarse mediante visualizaciones y dashboards interactivos que faciliten la interpretación de la información obtenida.

---------------------------------------------------------------------------

*`Objetivos Análiticos`*

- Analizar el comportamiento temporal de las ventas por producto y cliente.

- Identificar patrones, tendencias y anomalías en los datos históricos.

- Explorar la relación entre temporalidad, demanda e inventario.

- Aplicar medidas estadísticas y correlacionales para comprender el comportamiento de las variables.

- Detectar factores asociados a variaciones en la demanda de productos.

- Desarrollar modelos predictivos orientados a estimar ventas futuras.

- Identificar posibles riesgos de desabastecimiento en determinados periodos.

# 3. Importación de Librerias

En la sección actual se importarán todas las librerias necesarias para realizar todas las tareas desde el EDA hasta la realización de modelos de predicción.

In [1]:
#Si no se encuentra el paquete, se instala con el comando %pip
#%pip install scikit-learn
#%pip install seaborn

import pandas as pd
import numpy as np

# Visualización
import plotly.express as px
import plotly.graph_objects as go

# Machine Learning
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

# Estadística
from scipy.stats import pearsonr
import scipy.stats as stats
import seaborn as sns


# 4. Carga de Dataframes

Se cargan los dataframes.

In [ ]:
df_2023 = pd.read_csv("../data/2023.csv")
df_2024 = pd.read_csv("../data/2024.csv")
df_2025 = pd.read_csv("../data/2025.csv")

# 5. Comprensión Inicial de los Datos

En esta sección se realizará una exploración inicial del dataframe para comprender su estructura, dimensiones, tipos de datos y posibles desafíos asociados al análisis.

In [ ]:
print("Volumen 2023:")
print(df_2023.shape, "\n")
print("Volumen 2024:")
print(df_2024.shape, "\n")
print("Volumen 2025:")
print(df_2025.shape, "\n")

Volumen 2023:
(43338, 55) 

Volumen 2024:
(37035, 55) 

Volumen 2025:
(25239, 55) 



Se puede denotar que cada dataframe maneja una cantidad significativa de datos. Los tres dataframes manejan la misma cantidad de variables (columnas), sin embargo cada conjunto de datos tiene un número de instancias (filas) diferente. Curiosamente el número total de instancias es inversamente proporcional al pasar lo años, el 2023 siendo el mayor cuenta con 43338 instancias, el 2024 con 37025 y por ultimo el 2025 tiene 25239.

In [ ]:
print("COLUMNAS 2023")
print(df_2023.columns)

print("COLUMNAS 2024")
print(df_2024.columns)

print("COLUMNAS 2025")
print(df_2025.columns)

COLUMNAS 2023
Index(['Channel', 'Material Description', 'Category', '202301', '202302',
       '202303', '202304', '202305', '202306', '202307', '202308', '202309',
       '202310', '202311', '202312', '202313', '202314', '202315', '202316',
       '202317', '202318', '202319', '202320', '202321', '202322', '202323',
       '202324', '202325', '202326', '202327', '202328', '202329', '202330',
       '202331', '202332', '202333', '202334', '202335', '202336', '202337',
       '202338', '202339', '202340', '202341', '202342', '202343', '202344',
       '202345', '202346', '202347', '202348', '202349', '202350', '202351',
       '202352'],
      dtype='str')
COLUMNAS 2024
Index(['Channel', 'Material Description', 'Category', '202401', '202402',
       '202403', '202404', '202405', '202406', '202407', '202408', '202409',
       '202410', '202411', '202412', '202413', '202414', '202415', '202416',
       '202417', '202418', '202419', '202420', '202421', '202422', '202423',
       '202424', 

Los tres datasets presentan una estructura consistente con 55 columnas. Las primeras variables corresponden a información categórica del cliente, producto y categoría de negocio, mientras que las columnas restantes representan semanas del año en formato temporal (YYYYWW).


In [ ]:
print(df_2023.info())

<class 'pandas.DataFrame'>
RangeIndex: 43338 entries, 0 to 43337
Data columns (total 55 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   Channel               43338 non-null  str  
 1   Material Description  43338 non-null  str  
 2   Category              43338 non-null  str  
 3   202301                43338 non-null  int64
 4   202302                43338 non-null  int64
 5   202303                43338 non-null  int64
 6   202304                43338 non-null  int64
 7   202305                43338 non-null  int64
 8   202306                43338 non-null  int64
 9   202307                43338 non-null  int64
 10  202308                43338 non-null  int64
 11  202309                43338 non-null  int64
 12  202310                43338 non-null  int64
 13  202311                43338 non-null  int64
 14  202312                43338 non-null  int64
 15  202313                43338 non-null  int64
 16  202314         

El dataframe del 2023 cumple en que cada variable tiene el tipo de dato correspondiente.

In [ ]:
print(df_2024.info())

<class 'pandas.DataFrame'>
RangeIndex: 37035 entries, 0 to 37034
Data columns (total 55 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   Channel               37035 non-null  str  
 1   Material Description  37035 non-null  str  
 2   Category              37035 non-null  str  
 3   202401                37035 non-null  int64
 4   202402                37035 non-null  int64
 5   202403                37035 non-null  int64
 6   202404                37035 non-null  int64
 7   202405                37035 non-null  int64
 8   202406                37035 non-null  int64
 9   202407                37035 non-null  int64
 10  202408                37035 non-null  int64
 11  202409                37035 non-null  int64
 12  202410                37035 non-null  int64
 13  202411                37035 non-null  int64
 14  202412                37035 non-null  int64
 15  202413                37035 non-null  int64
 16  202414         

El dataframe del 2024 cumple en que cada variable tiene el tipo de dato correspondiente.

In [ ]:
print(df_2025.info())

<class 'pandas.DataFrame'>
RangeIndex: 25239 entries, 0 to 25238
Data columns (total 55 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Channel               25239 non-null  str    
 1   Material Description  25239 non-null  str    
 2   Category              25239 non-null  str    
 3   202501                25239 non-null  str    
 4   202502                25239 non-null  str    
 5   202503                25239 non-null  str    
 6   202504                25239 non-null  str    
 7   202505                25239 non-null  str    
 8   202506                25239 non-null  str    
 9   202507                25239 non-null  str    
 10  202508                25239 non-null  str    
 11  202509                25239 non-null  str    
 12  202510                25239 non-null  str    
 13  202511                25239 non-null  str    
 14  202512                25239 non-null  str    
 15  202513                25239 no

En cambio el 2025 presenta una anomalía con los tipos de datos, ya que bastantes semanas tienen tipo de cadena de texto (string), cuando estos representas las cantidades vendidas en dichas semanas.

Como los tres dataframes manejan la misma estructura se reconocen las siguientes variables:


- `CHANNEL`: Tipo texto Nominal, hace referencia al cliente, entidad o canal que esta realizando la compra de los productos.

- `MATERIAL DESCRIPTION`: Tipo texto Nominal, es el tipo de material o producto que se esta comprando.

- `CATEGORY`: Tipo texto Nominal, se refiere a la categoria en que se encuentra la compra. Sell in (Despacho a los clientes), Cust. Sales (Venta de los clientes del producto) y Channel Inv (Inventario).

- `202XN`: (X = 3 al 5) ( N = 01 al 52) Tipo Numérico Discreta. Representa la cantidad de unidades que se gestionaron en cada semana del año. En total son 52 columnas cada una representado cada semana del año.


In [ ]:
df_2023.head()

,Channel,Material Description,Category,202301,202302,202303,202304,202305,202306,202307,...,202343,202344,202345,202346,202347,202348,202349,202350,202351,202352
0,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Sell-in,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Cust. Sales,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Channel Inv.,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Sell-in,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Cust. Sales,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


El dataset combina variables categóricas con múltiples columnas semanales asociadas al comportamiento temporal de ventas e inventario. Debido a que la información temporal se encuentra distribuida en formato ancho, será necesaria una transformación a formato longitudinal para facilitar el análisis exploratorio y el modelado predictivo.

Las semanas se encuentran `representadas como columnas independientes`, por lo que se aplicará una transformación estructural hacia formato longitudinal para centralizar la variable temporal en una única columna.

Posterior a esto hacemos un análisis a la presencia de datos nulos, NA's, anomalos o duplicados.

In [ ]:
print("Datos nulos 2023:")
print(df_2023.isnull().sum().sum(),"\n")

print("Datos nulos 2024:")
print(df_2024.isnull().sum().sum(), "\n")

print("Datos nulos 2025:")
print(df_2025.isnull().sum().sum(), "\n")

Datos nulos 2023:
0 

Datos nulos 2024:
0 

Datos nulos 2025:
0 



Ninguno de los conjunto de datos presenta datos nulos

In [ ]:
print("Datos duplicados 2023:")
print(df_2023.duplicated().sum(), "\n")

print("Datos duplicados 2024:")
print(df_2024.duplicated().sum(), "\n")

print("Datos duplicados 2025:")
print(df_2025.duplicated().sum())

Datos duplicados 2023:
1741 

Datos duplicados 2024:
1620 

Datos duplicados 2025:
734


Cada año presenta una cantidad considerable de filas duplicadas, las cuales se trataran en la sección de transformación.

Posteriormente se revisa la cantidad de clientes/canales unicos que tiene cada dataframe. Adicionalmente tambien se ve la cantidad de productos unicos que se gestionaron en cada año.

In [ ]:
print("Total de clientes/canales unicos 2023:")
print(df_2023["Channel"].nunique())

print("Total de clientes/canales unicos 2024:")
print(df_2024["Channel"].nunique())

print("Total de clientes/canales unicos 2025:")
print(df_2025["Channel"].nunique())

Total de clientes/canales unicos 2023:
92
Total de clientes/canales unicos 2024:
80
Total de clientes/canales unicos 2025:
74


In [ ]:
print("Total de Productos unicos 2023:")
print(df_2023["Material Description"].nunique())

print("Total de Productos unicos 2024:")
print(df_2024["Material Description"].nunique())

print("Total de Productos unicos 2025:")
print(df_2025["Material Description"].nunique())

Total de Productos unicos 2023:
1645
Total de Productos unicos 2024:
1386
Total de Productos unicos 2025:
1159


Tras esto vamos revisar mas a profundidad la variable "Material Description", esto con el fin de desglosarla y obtener lo que esta descripción contiene a modo de obtener un mejor entendimiento de la variable, que se puede analizar de esta y si se puede analizar en partes.

In [ ]:
df_2023["Material Description"].str.split(",").str.len().value_counts()

Material Description
4    20955
5    13011
3     6978
6     2274
2       75
1       45
Name: count, dtype: int64

La descripción esta compuesta por palabras que se refieren a diferentes caracteristicas del producto, en efecto `cada caracteristica esta separada por una coma`, llegando a tener desde solo una caracteristica hasta otros a tener 6 caracteristicas.

Usando el ejemplo del producto `"MOBILE,SM-A032M/DS,BLACK,LTC"` se observa que este descripción esta organizado de la siguiente manera: tipo de producto, modelo, color y región. Pero como se observo anteriormente algunos productos tienen mas o menos caracteristicas que este ejemplo.

Por lo que se tomo cada año y se analizo las caracteristicas de cada producto en base al ejemplo y el orden que tenía, pero agregando 2 más que son el maximo, estos son: tipo de producto, modelo, variante, color, region y extra.

In [ ]:
split_cols = df_2023["Material Description"].str.split(",", expand=True)
split_cols.columns = [
    "Product_Type",
    "Model",
    "Variant",
    "Color",
    "Region",
    "Extra"
]
print("2023:")
for col in split_cols.columns:
    print(f"\n{col}")
    print(split_cols[col].value_counts().head())

2023:

Product_Type
Product_Type
MOBILE     14043
LED TV      6495
QLED TV     3489
TABLET      2472
OWM         1566
Name: count, dtype: int64

Model
Model
SELACOL        3753
26.5            558
SM-A135M/DS     492
SM-A032M/DS     417
SM-A127M/DS     399
Name: count, dtype: int64

Variant
Variant
BLACK        5028
CO           3738
120V 60HZ    3516
55           2379
KW           2286
Name: count, dtype: int64

Color
Color
COLOMBIA         10992
LTC              10062
COO               5154
UNITED STATES     1419
SETIN             1146
Name: count, dtype: int64

Region
Region
CO           1314
HP            960
TC            321
T             264
UWK10/U50     231
Name: count, dtype: int64

Extra
Extra
R410A    2268
R22         6
Name: count, dtype: int64


In [ ]:
split_cols = df_2024["Material Description"].str.split(",", expand=True)
split_cols.columns = [
    "Product_Type",
    "Model",
    "Variant",
    "Color",
    "Region",
    "Extra"
]
print("2024:")
for col in split_cols.columns:
    print(f"\n{col}")
    print(split_cols[col].value_counts().head())

2024:

Product_Type
Product_Type
MOBILE     11328
LED TV      4827
QLED TV     2754
TABLET      2157
OWM         1962
Name: count, dtype: int64

Model
Model
SELACOL        3390
SAMCOL          468
SM-S928B/DS     450
26.5            402
0.8             396
Name: count, dtype: int64

Variant
Variant
120V 60HZ    3900
CO           3807
BLACK        3555
KW           2193
GRAY         2043
Name: count, dtype: int64

Color
Color
LTC              9825
COLOMBIA         8469
COO              2580
UNITED STATES    1437
SETIN            1272
Name: count, dtype: int64

Region
Region
HP           1194
CO            999
TC            210
QWK30/Q50     198
QWK30/Q55     189
Name: count, dtype: int64

Extra
Extra
R410A    1875
R32       318
Name: count, dtype: int64


In [ ]:
split_cols = df_2025["Material Description"].str.split(",", expand=True)
split_cols.columns = [
    "Product_Type",
    "Model",
    "Variant",
    "Color",
    "Region",
    "Extra"
]
print("2025:")
for col in split_cols.columns:
    print(f"\n{col}")
    print(split_cols[col].value_counts().head())

2025:

Product_Type
Product_Type
MOBILE     7689
LED TV     3207
QLED TV    1863
TABLET     1422
OWM        1314
Name: count, dtype: int64

Model
Model
SELACOL        2208
SAMCOL          456
SM-A055M/DS     333
SM-S721B/DS     321
SM-S938B/DS     321
Name: count, dtype: int64

Variant
Variant
120V 60HZ    2733
CO           2625
BLACK        2340
KW           1563
GRAY         1428
Name: count, dtype: int64

Color
Color
LTC              6570
COLOMBIA         5685
COO              1752
UNITED STATES     993
SETIN             819
Name: count, dtype: int64

Region
Region
CO           795
HP           771
QWK30/Q50    147
UWM00/U40    147
UWS41/U50    147
Name: count, dtype: int64

Extra
Extra
R410A    1197
R32       369
Name: count, dtype: int64


Al analizar los resultados de la columna "Material Description" se puede concluir que el dataframe no sigue una ontología uniforme de producto, lo cual es completamente normal en datos empresariales.

Sin embargo en cada descripción se cumple algo, la primera caracteristica siempre corresponde al tipod de producto. Entonces conservarlo como una columna separada puede ser viable.

En la transformación se recomienda manejar un dataframe con una estructura con formato longitudinal, esto con el fin de normalizar y manejar una columna con la información de la semana y otra deñ año.

# 6. Integración y Transformación de Datos

En esta sección se realizan las transformaciones necesarias al conjunto de datos para que sea mas facil de trabajar, analizar y procesar.

Iniciamos con eliminar las `filas duplicadas`en cada dataframe.

In [ ]:
df_2023c = df_2023.drop_duplicates()
df_2024c = df_2024.drop_duplicates()
df_2025c = df_2025.drop_duplicates()

In [ ]:
print("Datos duplicados 2023:")
print(df_2023c.duplicated().sum(), "\n")

print("Datos duplicados 2024:")
print(df_2024c.duplicated().sum(), "\n")

print("Datos duplicados 2025:")
print(df_2025c.duplicated().sum())

Datos duplicados 2023:
0 

Datos duplicados 2024:
0 

Datos duplicados 2025:
0


Tras esto se corrige el tipo de dato de las columnas de las semanas del año 2025, teniendo en cuenta que tiene una estructura en que la "," representa el separador de miles y que tienen un espacio en blanco al final de algunos valores que pueden interferir con el parseo. Para finalmente transformar los datos de las columnas al formato numérico de forma segura.

In [ ]:
week_columns = df_2025c.columns[3:]

df_2025c[week_columns] = (
    df_2025c[week_columns]
    .replace(',', '', regex=True)
    .replace(' ', '', regex=True)
    .apply(pd.to_numeric, errors='coerce')
)

Al tener todos los datasets listos para trabajar creamos un método, el cual tiene el objetivo de que convierta las variables de las semanas en parte de las instancias y adicionalmente que cree una columna nueva nombrada "Units" que tome la cantidad de unidades que estaba en cada columna.

Esto aumentara significativamente el número de cada instancia, ya que si cada fila tenia información relacionada con cada columna de semana, ahora se transforma a que cada fila de dicho producto tenga la misma fila pero con una semana diferente.

In [ ]:
def transform_date(df):
    df = df.melt(id_vars=["Channel", "Material Description", "Category"], var_name="Semana", value_name="Units")
    df["Year"] = df["Semana"].str[:4]
    df["Year"] = df["Year"].astype(int)
    df["Week"] = df["Semana"].str[4:]
    df["Week"] = df["Week"].astype(int)
    df["Units"] = df["Units"].astype(int)
    df = df.drop("Semana", axis=1)
    return df

In [ ]:
df_2023t = transform_date(df_2023c)
df_2024t = transform_date(df_2024c)
df_2025t = transform_date(df_2025c)
df_2023t.head()

,Channel,Material Description,Category,Units,Year,Week
0,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Sell-in,0,2023,1
1,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Cust. Sales,0,2023,1
2,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Channel Inv.,0,2023,1
3,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Sell-in,0,2023,1
4,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Cust. Sales,0,2023,1


Los valores iguales a cero fueron conservados debido a que representan semanas sin movimiento comercial, comportamiento esperado en series temporales de ventas e inventario.

Tras realizar la transformación mencionada se observa que ahora muestran los productos en orden pero desde la semana uno del año vigente. Pero esto tambien puede generar datos duplicados los cuales tambien se trataran inmediatamente en la siguiente celda.

In [ ]:
print("Datos duplicados 2023 transformado:")
print(df_2023t.duplicated().sum(), "\n")
df_2023t = df_2023t.drop_duplicates()

print("Datos duplicados 2024 transformado:")
print(df_2024t.duplicated().sum(), "\n")
df_2024t = df_2024t.drop_duplicates()

print("Datos duplicados 2025 transformado:")
print(df_2025t.duplicated().sum())
df_2025t = df_2025t.drop_duplicates()

Datos duplicados 2023 transformado:
139783 

Datos duplicados 2024 transformado:
141318 

Datos duplicados 2025 transformado:
113736


Tras trasnformar cada dataframe de forma separada se puede proceder a unirlos. Para posteriormente ver si tienen datos nulos o duplicados, asi como ver la información de las variables y el número de instancias.

In [ ]:
df_total = pd.concat([df_2023t, df_2024t, df_2025t], ignore_index=True)
print(df_total.info())
print("Total de datos nulos: ", df_total.isnull().sum().sum())
print("Total de datos duplicados : ", df_total.duplicated().sum())

<class 'pandas.DataFrame'>
RangeIndex: 4884047 entries, 0 to 4884046
Data columns (total 6 columns):
 #   Column                Dtype
---  ------                -----
 0   Channel               str  
 1   Material Description  str  
 2   Category              str  
 3   Units                 int64
 4   Year                  int64
 5   Week                  int64
dtypes: int64(3), str(3)
memory usage: 223.6 MB
None
Total de datos nulos:  0
Total de datos duplicados :  0


In [ ]:
df_total.head()

,Channel,Material Description,Category,Units,Year,Week
0,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Sell-in,0,2023,1
1,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Cust. Sales,0,2023,1
2,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Channel Inv.,0,2023,1
3,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Sell-in,0,2023,1
4,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Cust. Sales,0,2023,1


In [ ]:
df_total.tail()

,Channel,Material Description,Category,Units,Year,Week
4884042,CUSTOMER90,"MOBILE,SM-S938B/DS,BLACK,LTC",Cust. Sales,0,2025,52
4884043,CUSTOMER90,"MOBILE,SM-S938B/DS,BLACK,LTC",Channel Inv.,0,2025,52
4884044,CUSTOMER90,"TABLET,SM-X205,GRAY,COO",Sell-in,0,2025,52
4884045,CUSTOMER90,"TABLET,SM-X205,GRAY,COO",Cust. Sales,0,2025,52
4884046,CUSTOMER90,"TABLET,SM-X205,GRAY,COO",Channel Inv.,0,2025,52


Tras concatenar los dataframes de cada año en un datafrane general, se puede evidenciar que se unieron sin ningun problema, adicionalmente que los columnas mantienen el tipo de dato adecuado, esto tambien se ve al obtener algunas filas del dataframe.

Tras combinar los dataframes, ahora se puede extraer el tipo de producto de la descripción del material.

In [ ]:
df_total['Product_Type'] = (df_total['Material Description'].str.split(',').str[0])
df_total

,Channel,Material Description,Category,Units,Year,Week,Product_Type
0,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Sell-in,0,2023,1,MOBILE
1,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Cust. Sales,0,2023,1,MOBILE
2,CUSTOMER1,"MOBILE,SM-A032M/DS,BLACK,LTC",Channel Inv.,0,2023,1,MOBILE
3,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Sell-in,0,2023,1,MOBILE
4,CUSTOMER1,"MOBILE,SM-A035M/DS,BLUE,LTC",Cust. Sales,0,2023,1,MOBILE
...,...,...,...,...,...,...,...
4884042,CUSTOMER90,"MOBILE,SM-S938B/DS,BLACK,LTC",Cust. Sales,0,2025,52,MOBILE
4884043,CUSTOMER90,"MOBILE,SM-S938B/DS,BLACK,LTC",Channel Inv.,0,2025,52,MOBILE
4884044,CUSTOMER90,"TABLET,SM-X205,GRAY,COO",Sell-in,0,2025,52,TABLET
4884045,CUSTOMER90,"TABLET,SM-X205,GRAY,COO",Cust. Sales,0,2025,52,TABLET


`La variable Material Description fue descompuesta para mejorar la interpretabilidad y el análisis segmentado de productos.`

Se crea una nueva columna "Produc Type", que viene de la primera caracteristica de "Material Description". Ahora se puede organizar un poco el dataframe para que se vea mas organizado el orden las variables.

In [ ]:
df_total = df_total.reindex(columns=['Channel',  'Product_Type', 'Material Description', 'Category', 'Week', 'Year', 'Units'])
df_total.head()

,Channel,Product_Type,Material Description,Category,Week,Year,Units
0,CUSTOMER1,MOBILE,"MOBILE,SM-A032M/DS,BLACK,LTC",Sell-in,1,2023,0
1,CUSTOMER1,MOBILE,"MOBILE,SM-A032M/DS,BLACK,LTC",Cust. Sales,1,2023,0
2,CUSTOMER1,MOBILE,"MOBILE,SM-A032M/DS,BLACK,LTC",Channel Inv.,1,2023,0
3,CUSTOMER1,MOBILE,"MOBILE,SM-A035M/DS,BLUE,LTC",Sell-in,1,2023,0
4,CUSTOMER1,MOBILE,"MOBILE,SM-A035M/DS,BLUE,LTC",Cust. Sales,1,2023,0


In [ ]:
print(df_total.info())

<class 'pandas.DataFrame'>
RangeIndex: 4884047 entries, 0 to 4884046
Data columns (total 7 columns):
 #   Column                Dtype 
---  ------                ----- 
 0   Channel               str   
 1   Product_Type          object
 2   Material Description  str   
 3   Category              str   
 4   Week                  int64 
 5   Year                  int64 
 6   Units                 int64 
dtypes: int64(3), object(1), str(3)
memory usage: 260.8+ MB
None


In [ ]:
print("\n\n", "Datos nulos:", "\n")
print(df_total.isnull().sum())
print("\n\n", "Datos dupĺicados:")
print(df_total.duplicated().sum())



 Datos nulos: 

Channel                 0
Product_Type            0
Material Description    0
Category                0
Week                    0
Year                    0
Units                   0
dtype: int64


 Datos dupĺicados:
0


Al terminar con el tratamiento de los dataframes, las transformaciones, la unión y su gestión se obtiene un dataframe con 7 variables que corresponden a:


- `CHANNEL`: Tipo texto Nominal, hace referencia al cliente, entidad o canal que esta realizando la compra de los productos.
  
- `PRODUCT_TYPE`: Tipo texto Nominal, es el tipo de material o producto que se esta comprando.

- `MATERIAL DESCRIPTION`: Tipo texto Nominal, es una linea de texto que contiene varias caracteristicas del producto.

- `CATEGORY`: Tipo texto Nominal, se refiere a la categoria en que se encuentra la compra. Sell in (Despacho a los clientes), Cust. Sales (Venta de los clientes del producto) y Channel Inv (Inventario).

- `WEEK`: Tipo Numérico Discreta. Representa la semana del año en que se realizo la compra. Va desde la semana 1 hasta la 52 de su respectivo año.

- `YEAR`: Tipo Numérico Discreta. Representa el año en que se realizo la compra. Va desde el 2023 hasta el 2025.
  
- `UNITS`: Tipo Numérico Discreta. Es la cantidad de unidades que se gestionaron de cierto producto en cierta semana del año.

Tras terminar con esta fase se guardara el dataframe obtenido para utilizar este como insumo principal de aqui en adelante en las siguientes fases.

In [ ]:
df_total.to_csv("../data/dataset_transformado.csv", index=False)

# 7. Limpieza y Preprocesamiento

Aqui se tratan los datos nulos, valores atipicos, tipos de datos, inconsistencias y normalización si aplica para que no afecte el análisis más adelante.

In [2]:
## ADVERTENCIA: Comentar esta celda si NO está trabajando dentro de Google Colaboratory
import os
from google.colab import drive
drive.mount('/content/drive')  # AVISO: conectar desde el dueño directo del drive (quien tenga MyDrive)

os.getcwd()
os.chdir("/content/drive/MyDrive")  # AVISO: no conecta bien a Shared with me
os.getcwd()

Mounted at /content/drive


'/content/drive/MyDrive'

In [22]:
df_total = pd.read_csv('dataset_transformado.csv', header='infer')

### 7.1 Carga del Dataset Transformado

Debido al tamaño del dataframe obtenido al final de la transformación (más de 4.8 millones de registros), se alamacenó una versión procesada del dataframe como ya se menciono con el fin de optimizar tiempos de ejecución y facilitar las etapas posteriores del análisis.

In [ ]:
df_total = pd.read_csv("../data/dataset_transformado.csv")

### 7.2 Verificación General del Dataset

Tras importar el archivo generado en la transformación se revisará la calidad de los datos. Desde la presencia de nulos y duplicados, hasta los outliers presentes en el datos y que los tipos de datos correspondan a la variable adecuada.

In [ ]:
print("\n\n", "Datos nulos:", "\n")
print(df_total.isnull().sum())



 Datos nulos: 

Channel                 0
Product_Type            0
Material Description    0
Category                0
Week                    0
Year                    0
Units                   0
dtype: int64


No se presenta ningún dato nulo o Nan en el conjunto.

In [ ]:
print("Datos dupĺicados:")
print(df_total.duplicated().sum())

Datos dupĺicados:
0


Tampoco se tiene  la presencia de datos o filas duplicadas en el dataframe

In [ ]:
print(df_total.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4884047 entries, 0 to 4884046
Data columns (total 7 columns):
 #   Column                Dtype 
---  ------                ----- 
 0   Channel               object
 1   Product_Type          object
 2   Material Description  object
 3   Category              object
 4   Week                  int64 
 5   Year                  int64 
 6   Units                 int64 
dtypes: int64(3), object(4)
memory usage: 260.8+ MB
None


No se presenta ninguna discordancia en las variables del dataframe con sus tipo de datos correspondientes.

Sin embargo, la variable `Units` presenta valores negativos y registros extremadamente altos, por lo que será necesario analizar su distribución y determinar si corresponden a comportamientos reales del negocio o posibles anomalías.

### 7.3 Validación de Variables Categóricas

Se verfifica que solo existan categorías validas y que no haya inconsistencias semánticas con las categorias.

In [ ]:
print(df_total["Category"].value_counts())

Category
Channel Inv.    1685110
Cust. Sales     1630016
Sell-in         1568921
Name: count, dtype: int64


### 7.4 Análisis de Distribución de Units

#### 7.4.1 Estadísticos descriptivos

En este apartado se analizará estadisticamente la variable "Units" al ser una de nuestras variables objetivo en análisis, modelado y creación del modelo.

In [ ]:
print(df_total["Units"].describe())

count    4.884047e+06
mean     1.902829e+01
std      2.125423e+02
min     -3.997000e+03
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      4.348400e+04
Name: Units, dtype: float64


In [ ]:
print(df_total["Units"].describe(percentiles=[0.90,0.95,0.99,0.999]))

count    4.884047e+06
mean     1.902829e+01
std      2.125423e+02
min     -3.997000e+03
50%      0.000000e+00
90%      1.100000e+01
95%      4.200000e+01
99%      3.520000e+02
99.9%    2.521000e+03
max      4.348400e+04
Name: Units, dtype: float64


In [ ]:
print(df_total["Units"].describe(percentiles=[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]))

count    4.884047e+06
mean     1.902829e+01
std      2.125423e+02
min     -3.997000e+03
10%      0.000000e+00
20%      0.000000e+00
30%      0.000000e+00
40%      0.000000e+00
50%      0.000000e+00
60%      0.000000e+00
70%      0.000000e+00
80%      1.000000e+00
90%      1.100000e+01
max      4.348400e+04
Name: Units, dtype: float64


Los estadísticos descriptivos evidencian una distribución altamente dispersa y asimétrica. La presencia masiva de valores iguales a cero genera un comportamiento sparse, mientras que algunos registros extremos elevan considerablemente la desviación estándar y el rango de la variable.

#### 7.4.2 Análisis de Sparsity (Ceros)

In [ ]:
df_total["Units"].value_counts().head(10)

,count
Units,
0,3704520
1,227442
2,115750
3,77491
4,58403
5,53979
6,36226
7,29846
10,27704


In [ ]:
zero_pct = (df_total["Units"] == 0).mean() * 100
total = len(df_total)

print(f"Porcentaje de ceros: {zero_pct:.2f}%")

Porcentaje de ceros: 75.85%


In [ ]:
non_zero_pct = (df_total["Units"] != 0).mean() * 100
print(f"Porcentaje de NO ceros: {non_zero_pct:.2f}%")

Porcentaje de NO ceros: 24.15%


$75.85 \%  \text{ de los datos son 0}$

Al obtener la cantidad de 0's que son 3704520, equivale a mas del 75% por ciento del total de 4.88 millones de registros. Analizandolo detenidamente esta cantidad de 0's no son un problema ya que pueden representar que no hubo movimiento como en los siguientes casos:
- Semanas sin ventas.
- Productos inactivos.
- Productos sin movimiento.
- Inventario sin rotación.
- Clientes que no compraron cierto producto esa semana.


Debido al contexto transaccional del negocio, estos valores pueden representar semanas sin ventas, productos sin rotación o ausencia de movimiento comercial. Por esta razón, los registros iguales a cero serán conservados para las etapas posteriores del análisis.

#### 7.4.3 Análisis de Valores Negativos

In [ ]:
print("Total de Negativos : ", (df_total["Units"] < 0).sum())


Total de Negativos :  13373


In [ ]:
print(df_total[df_total["Units"] < 0]["Category"].value_counts())

Category
Sell-in         4928
Cust. Sales     4610
Channel Inv.    3835
Name: count, dtype: int64


In [ ]:
negative_pct = (df_total["Units"] < 0).mean() * 100
print(f"Porcentaje de Negativos: {negative_pct:.2f}%")

Porcentaje de Negativos: 0.27%


In [ ]:
print(df_total[df_total["Units"]<-500].head(20))

            Channel Product_Type                      Material Description  \
109176   CUSTOMER70       MOBILE         MOBILE,SM-A135M/DS,LIGHT BLUE,LTC   
109182   CUSTOMER70       MOBILE              MOBILE,SM-A135M/DS,BLACK,LTC   
259846   CUSTOMER57       MOBILE              MOBILE,SM-A037M/DS,BLACK,LTC   
259880   CUSTOMER57       MOBILE              MOBILE,SM-A135M/DS,BLACK,LTC   
1434424  CUSTOMER73       MOBILE       MOBILE,SM-A145M/DS,LIGHT GREEN ,LTC   
1434427  CUSTOMER73       MOBILE              MOBILE,SM-A145M/DS,BLACK,LTC   
1614325  CUSTOMER43  AV RECEIVER              AV RECEIVER,HW-C450,COLOMBIA   
1653825  CUSTOMER43       MOBILE       MOBILE,SM-A145M/DS,LIGHT GREEN ,LTC   
1653828  CUSTOMER43       MOBILE              MOBILE,SM-A145M/DS,BLACK,LTC   
1653831  CUSTOMER43       MOBILE             MOBILE,SM-A145M/DS,SILVER,LTC   
2541097  CUSTOMER73       MOBILE              MOBILE,SM-A055M/DS,BLACK,LTC   
3405472  CUSTOMER20       MOBILE                 MOBILE,SM-A057M

Los valores negativos representan únicamente el $0.27\%$ del total de registros. Debido al contexto operacional del negocio, estos valores podrían estar asociados a devoluciones, ajustes de inventario, cancelaciones o correcciones contables.

Dado su posible significado analítico y su baja proporción dentro del dataset, estos registros serán conservados para etapas posteriores del análisis y modelado.

### 7.5 Distribución de Movimientos Reales

In [ ]:
sample_units = df_total["Units"].sample(100000, random_state=69)

Debido al tamaño del dataset, se utilizará una muestra aleatoria de 100000 registros con el fin de facilitar la visualización y reducir el consumo computacional, manteniendo una representación adecuada del comportamiento general de los datos.

In [ ]:
golden_number = (1 + np.sqrt(5))/2
fig = px.histogram( sample_units
               , width = 900, height=900/golden_number
               , nbins = 100
               , marginal= 'rug'
               , orientation="v"
               ,color_discrete_sequence =["purple"]
               ,title = "Distribución de Units (Muestra de 100k registros)"
                )

fig.update_layout(bargap=0.05)


fig.update_traces(selector=dict(type='histogram'),overwrite=True, opacity=0.6
                  ,  xbins = dict(start = min(sample_units), end=max(sample_units), size =   np.ptp(np.histogram_bin_edges(sample_units, bins='auto')[0:2]))
                  , autobinx=False
                  )
fig.show()

La distribución inicial evidencia una fuerte concentración de registros cercanos a cero, lo que dificulta visualizar adecuadamente el comportamiento de los movimientos reales de ventas e inventario.

In [8]:
df_nonzero = df_total[df_total["Units"] > 0]
sample_units = df_nonzero["Units"].sample(100000, random_state=69)

In [ ]:
golden_number = (1 + np.sqrt(5))/2
fig = px.histogram( sample_units
               , width = 900, height=900/golden_number
               , nbins = 100
               , marginal= 'rug'
               , orientation="v"
               ,color_discrete_sequence =["purple"]
               ,title = "Distribución de Units (>0)"
                )

fig.update_layout(bargap=0.05)


fig.update_traces(selector=dict(type='histogram'),overwrite=True, opacity=0.6
                  ,  xbins = dict(start = min(sample_units), end=max(sample_units), size =   np.ptp(np.histogram_bin_edges(sample_units, bins='auto')[0:2]))
                  , autobinx=False
                  )
fig.show()

Debido a la alta asimetría de la variable, se aplicará una transformación logarítmica sobre los valores positivos con el objetivo de comprimir el rango de los datos y visualizar con mayor claridad la distribución de los movimientos comerciales.

In [11]:
golden_number = (1 + np.sqrt(5))/2
sample_log = np.log1p(sample_units)

fig = px.histogram( sample_log
               , width = 900, height=900/golden_number
               , nbins = 101
               , marginal= 'rug'
               , orientation="v"
               ,color_discrete_sequence =["purple"]
               ,title = "Distribución Logarítmica de Units"
                )

fig.update_layout(bargap=0.05)


fig.update_traces(selector=dict(type='histogram'),overwrite=True, opacity=0.6
                  ,  xbins = dict(start = min(sample_log), end=max(sample_log), size =  np.ptp(np.histogram_bin_edges(sample_log, bins='auto')[0:2]))
                  , autobinx=False
                  )
fig.show()


Output hidden; open in https://colab.research.google.com to view.

In [15]:
fig = px.box(
    y=sample_log,
    title="Boxplot Logarítmico de Units"
)

fig.show()

La variable `Units` presenta una distribución fuertemente asimétrica positiva, caracterizada por una alta concentración de registros en valores bajos o iguales a cero y una cola extensa hacia la derecha compuesta por pocos valores extremos.

Los percentiles muestran que incluso el 75% de los registros permanecen cercanos a cero, evidenciando un comportamiento sparse típico en datos transaccionales de ventas e inventario.

El histograma y la transformación logarítmica permiten observar con mayor claridad la distribución de los movimientos comerciales positivos, donde la mayoría de operaciones corresponden a volúmenes pequeños y solo un reducido grupo concentra cantidades considerablemente altas.

Adicionalmente, se identifican registros negativos en una proporción mínima, los cuales podrían estar asociados a devoluciones, ajustes operativos o correcciones de inventario dentro de la operación comercial.

### 7.7 Tratamiento de Outliers

In [ ]:
print("Cantidad de Units < 0: ", df_total[df_total["Units"]  < 0].value_counts().sum())
print("Cantidad de Units 0: ", df_total[df_total["Units"]  == 0].value_counts().sum())
print("Cantidad de Units 1 -> 1000: ", df_total[(df_total["Units"]  >= 1) & (df_total["Units"]  < 1000)]["Units"].value_counts().sum())
print("Cantidad de Units 1000 -> 5000: ", df_total[(df_total["Units"]  >= 1000) & (df_total["Units"]  < 5000)]["Units"].value_counts().sum())
print("Cantidad de Units 5000 -> 10000: ", df_total[(df_total["Units"]  >= 5000) & (df_total["Units"]  < 10000)]["Units"].value_counts().sum())
print("Cantidad de Units 10000 -> 20000: ", df_total[(df_total["Units"]  >= 10000) & (df_total["Units"]  < 20000)]["Units"].value_counts().sum())
print("Cantidad de Units 20000 -> 30000: ", df_total[(df_total["Units"]  >= 20000) & (df_total["Units"]  < 30000)]["Units"].value_counts().sum())
print("Cantidad de Units > 30000: ", df_total[df_total["Units"]  >= 30000].value_counts().sum())

Cantidad de Units < 0:  13373
Cantidad de Units 0:  3704520
Cantidad de Units 1 -> 1000:  1147826
Cantidad de Units 1000 -> 5000:  16881
Cantidad de Units 5000 -> 10000:  1068
Cantidad de Units 10000 -> 20000:  349
Cantidad de Units 20000 -> 30000:  24
Cantidad de Units > 30000:  6


In [7]:
aux = {
    'Units value': ['0', 'Negativos', 'Positvos'],
    'Units pct': [((df_total["Units"] == 0).mean() * 100), ((df_total["Units"] < 0).mean() * 100), ((df_total["Units"] > 0).mean() * 100)]
}
fig = px.pie(aux, values='Units pct', names='Units value', title='Proporciones de Units')
aux = pd.DataFrame(aux)
print(aux)
fig.show()


  Units value  Units pct
0           0  75.849393
1   Negativos   0.273810
2    Positvos  23.876797


Analizando el número de valores en Units la mayor proporción esta en 0's como ya se menciono, sin embargo se denota que sin mencionar lo de 0 la mayoria de clientes hacia movimientos de 1 a 1000 productos. Por otro lado algunos clientes manejaban grandes cifras de productos, alrededor de 18000 mil registros se movieron mas de 1000 productos, pero entre mas cantidad movian menos se repetia el movimiento, llegando a presentarse 6 movimientos de más de 30000 mil productos. Ventas así pueden llegar a presentarse, asi como inventarios enormes tambien.

<br>

Debido a la naturaleza comercial del problema, no se aplicará eliminación ni truncamiento agresivo de outliers durante esta etapa, ya que estos valores podrían representar eventos reales de alta demanda o movimientos excepcionales relevantes para el análisis predictivo.

### 7.8 Conlusiones de Limpieza


Tras finalizar con esta etapa se afirma que se tiene un conjunto de datos consistente, el cual se transformo de una manera que no presenta datos nulos ni duplicados. Al análizar detenidamente la distribución de la variable objetivo "Units" sale a relucir la cantidad alta que se presenta, lo que puede representar que no hubo movimiento de dicho producto en esa semana.

<br>

Cabe mencionar la presencia de valores negativos en la variable, pero al no representar mas del 1% del total de los datos tienen un alto valor analítico al poder relacionarlos con devoluciones, ajustes o correcciones con respecto a los productos, siendo esta la razón para conservar y evaluar estos mas adelante. Asi como existe datos negativos existen datos que estan bastante alejados del resto de registros, llegandose a presentar movimiento de mas de 30000 productos, sin embargo igual que al caso anterior no llegan a representar mas del 1% de los datos y estas ventas pueden ser posibles y vale la pena analizarlos sin alterarlos.

<br>

Tras el análisis de calidad de datos se concluye que el conjunto no presenta valores nulos ni duplicados luego de la transformación.
Los valores negativos fueron conservados debido a que pueden representar devoluciones, ajustes operativos o correcciones de inventario.
Los valores iguales a cero también fueron mantenidos, ya que reflejan semanas sin movimiento comercial y forman parte del comportamiento real de la demanda.
Los valores extremos altos no fueron eliminados debido a que pueden corresponder a temporadas de alta rotación o eventos comerciales relevantes.
Por lo tanto, no se aplicó eliminación agresiva de registros para evitar pérdida de información de negocio.

# 8. Ingeniería de Características (Feature Engineering)

Aqui se puede crear intervalos temporales, como trimestres, semestres. O incluso variables calculadas que tienen infomación valiosa como promedio móvil, crecimiento porcentual, las ventas acumuladas o las tendencias temporales.

#### 8.1 Objetivo del Feature Engineering

En esta etapa se construirán nuevas variables a partir de los datos históricos con el objetivo de capturar patrones temporales, tendencias de demanda y comportamiento comercial. Esta fase busca transformar los datos originales en variables como:

* Representen mejor el comportamiento del negocio.
* Ayuden al modelo a detectar patrones.
* Mejoren capacidad predictiva.
* Capturen temporalidad.
* Capturen demanda.
* Capturen estacionalidad.
* Capturen comportamiento histórico.

Estas características permitirán enriquecer el análisis exploratorio y mejorar el rendimiento de futuros modelos predictivos relacionados con ventas e inventario.

#### 8.2 Variables Temporales

In [23]:
df_total["Date"] = pd.to_datetime(
    df_total["Year"].astype(str) + "-" +
    df_total["Week"].astype(str) + "-1",
    format="%Y-%W-%w"
)

In [24]:
df_total.tail()

,Channel,Product_Type,Material Description,Category,Week,Year,Units,Date
4884042,CUSTOMER90,MOBILE,"MOBILE,SM-S938B/DS,BLACK,LTC",Cust. Sales,52,2025,0,2025-12-29
4884043,CUSTOMER90,MOBILE,"MOBILE,SM-S938B/DS,BLACK,LTC",Channel Inv.,52,2025,0,2025-12-29
4884044,CUSTOMER90,TABLET,"TABLET,SM-X205,GRAY,COO",Sell-in,52,2025,0,2025-12-29
4884045,CUSTOMER90,TABLET,"TABLET,SM-X205,GRAY,COO",Cust. Sales,52,2025,0,2025-12-29
4884046,CUSTOMER90,TABLET,"TABLET,SM-X205,GRAY,COO",Channel Inv.,52,2025,0,2025-12-29


In [25]:
df_total["Month"] = df_total["Date"].dt.month
df_total["Quarter"] = df_total["Date"].dt.quarter
df_total["YearMonth"] = df_total["Date"].dt.to_period("M").astype(str)

In [26]:
df_total.tail()

,Channel,Product_Type,Material Description,Category,Week,Year,Units,Date,Month,Quarter,YearMonth
4884042,CUSTOMER90,MOBILE,"MOBILE,SM-S938B/DS,BLACK,LTC",Cust. Sales,52,2025,0,2025-12-29,12,4,2025-12
4884043,CUSTOMER90,MOBILE,"MOBILE,SM-S938B/DS,BLACK,LTC",Channel Inv.,52,2025,0,2025-12-29,12,4,2025-12
4884044,CUSTOMER90,TABLET,"TABLET,SM-X205,GRAY,COO",Sell-in,52,2025,0,2025-12-29,12,4,2025-12
4884045,CUSTOMER90,TABLET,"TABLET,SM-X205,GRAY,COO",Cust. Sales,52,2025,0,2025-12-29,12,4,2025-12
4884046,CUSTOMER90,TABLET,"TABLET,SM-X205,GRAY,COO",Channel Inv.,52,2025,0,2025-12-29,12,4,2025-12


#### 8.3 Variables de Movimiento Comercial

In [28]:
df_total["Positive_Movement"] = (
    df_total["Units"] > 0
).astype(int)

df_total["Negative_Movement"] = (
    df_total["Units"] < 0
).astype(int)

In [31]:
df_total.head()

,Channel,Product_Type,Material Description,Category,Week,Year,Units,Date,Month,Quarter,YearMonth,Positive_Movement,Negative_Movement
0,CUSTOMER1,MOBILE,"MOBILE,SM-A032M/DS,BLACK,LTC",Sell-in,1,2023,0,2023-01-02,1,1,2023-01,0,0
1,CUSTOMER1,MOBILE,"MOBILE,SM-A032M/DS,BLACK,LTC",Cust. Sales,1,2023,0,2023-01-02,1,1,2023-01,0,0
2,CUSTOMER1,MOBILE,"MOBILE,SM-A032M/DS,BLACK,LTC",Channel Inv.,1,2023,0,2023-01-02,1,1,2023-01,0,0
3,CUSTOMER1,MOBILE,"MOBILE,SM-A035M/DS,BLUE,LTC",Sell-in,1,2023,0,2023-01-02,1,1,2023-01,0,0
4,CUSTOMER1,MOBILE,"MOBILE,SM-A035M/DS,BLUE,LTC",Cust. Sales,1,2023,0,2023-01-02,1,1,2023-01,0,0


# 9. EDA - Análisis Exploratorio de Datos

Aquí se procede a realizar el análisis de variables y tendencias mediante gráficas. Esto con el fin de obtener información valiosa como la temporalidad, anomalías, top productos, etc.

# 10. Estadística e Hipótesis

Aqui corresponde a la sección donde se calcula medidas de tendencia central, dispersión, correlación y pruebas estadísticas, con el fin de obtener la validación o refutación de las hipótesis.

# 11. Modelado Predictivo

Procedemos con la sección donde creamos, entrenamos y evaluamos el modelo.

# 12. Evaluación del Modelo

Tras obtener el modelo lo evaluamos mediante los diferentes KPI's.

# 13. Clustering

Como análisis complementario segmentan los clientes y productos. Para obtener datos segun los agrupamientos.

# 14. Insights de Negocio

Entrando en la etapa final, aqui se responde segun lo aprendido de las demas etapas a los objetivos y problemas del negocio. Con esto se pueden empezar a tomar decisiones.

# 15. Respuesastas Estratégicas del Negocio

# 15. Conclusiones y Trabajo Futuro